test-data/unit/check-custom-plugin.test

```diff
@@ -229,3 +229,50 @@ def decorator2() -> Callable[..., Callable[..., int]]: pass
 [file mypy.ini]
 [[mypy]
 plugins=<ROOT>/test-data/unit/plugins/named_callable.py
+
+[case testMethodSignatureHook]
+# flags: --config-file tmp/mypy.ini
+from typing import Iterator
+
+class Foo:
+    # Test that method signature hooks are applied in various cases: explicit method calls, and
+    # implicit dunder method calls through language syntax.
+    # The plugin's method signature hook should turn all str occurrences into int.
+    def __init__(self, index: str) -> None: ...
+    def __getitem__(self, index: str) -> str: ...
+    def __setitem__(self, index: str, value: str) -> None: ...
+    def __iter__(self) -> Iterator[str]: ...
+    def __next__(self) -> str: ...
+    def m(self, arg: str) -> str: ...
+
+foo = Foo(1)
+reveal_type(foo.m(2)) # E: Revealed type is 'builtins.int'
+reveal_type(foo[3]) # E: Revealed type is 'builtins.int'
+foo[4] = 5
+for x in foo:
+    reveal_type(x) # E: Revealed type is 'builtins.int*'
+
+[file mypy.ini]
+[[mypy]
+plugins=<ROOT>/test-data/unit/plugins/method_sig_hook.py
+
+[case testMethodSignatureHookNamesFullyQualified]
+# flags: --config-file tmp/mypy.ini
+from mypy_extensions import TypedDict
+from typing import Any, NamedTuple
+
+class FullyQualifiedTestClass:
+    def instance_method(self) -> str: ...
+
+class FullyQualifiedTestTypedDict(TypedDict):
+    foo: str
+
+FullyQualifiedTestNamedTuple = NamedTuple('FullyQualifiedTestNamedTuple', [('foo', str)])
+
+# Check the return types to ensure that the method signature hook is called in each case
+reveal_type(FullyQualifiedTestClass().instance_method()) # E: Revealed type is 'builtins.int'
+reveal_type(FullyQualifiedTestNamedTuple('')._asdict()) # E: Revealed type is 'builtins.int'
+
+[file mypy.ini]
+[[mypy]
+plugins=<ROOT>/test-data/unit/plugins/fully_qualified_test_hook.py
```

Comment:
I think it would be great to add a test for class objects (assuming the fix I suggested works).

Reply:
Right, I actually had a test case for that, but removed it while testing. I'll re-add it once I get method signature hooks working on class methods.

Reply:
Something else I forgot to mention: I also wanted to add a test case for method calls on `TypedDict`s (that's why the `FullyQualifiedTestTypedDict` declaration is there). But it seems that mypy doesn't allow *any* method calls on `TypedDict` instances, even standard methods like `copy`. Am I missing something here?

If method calls on `TypedDict`s are really not allowed by mypy, it doesn't make much sense to support them in `method_fullname` (or at least there should be a comment explaining that the `TypedDict` branch is never called at the moment).